lib/SILOptimizer/FunctionSignatureTransforms/ExistentialTransform.cpp

```diff
@@ -0,0 +1,595 @@
+      auto OpenedArchetype = ContextTy->getAs<ArchetypeType>();
+      for (auto proto : OpenedArchetype->getConformsTo()) {
+        NewConformances.push_back(ProtocolConformanceRef(proto));
+      }
+      ArrayRef<ProtocolConformanceRef> Conformances =
+          Ctx.AllocateCopy(NewConformances);
+      auto ExistentialRepr =
+          ArgDesc.Arg->getType().getPreferredExistentialRepresentation(M);
+      switch (ExistentialRepr) {
+      case ExistentialRepresentation::Opaque: {
+        /// Create this sequence for init_existential_addr.:
+        /// bb0(%0 : $*T):
+        /// %3 = alloc_stack $P
+        /// %4 = init_existential_addr %3 : $*P, $T
+        /// copy_addr %0 to [initialization] %4 : $*T
+        /// destroy_addr %0 : $*T
+        /// %7 = open_existential_addr immutable_access %3 : $*P to
+        /// $*@opened P
+        auto *ASI =
+            NewFBuilder.createAllocStack(InsertLoc, ArgDesc.Arg->getType());
+        Arg2AllocStackMap.insert(
+            std::pair<int, SILInstruction *>(ArgDesc.Index, ASI));
+
+        auto *EAI = NewFBuilder.createInitExistentialAddr(
+            InsertLoc, ASI, NewArg->getType().getASTType()->getCanonicalType(),
+            NewArg->getType(), Conformances);
+        /// If DestroyAddr is already there, then do not use [take].
+        NewFBuilder.createCopyAddr(
+            InsertLoc, NewArg, EAI,
+            ExistentialArgDescriptor[ArgDesc.Index].AccessType ==
+                    OpenedExistentialAccess::Immutable
+                ? IsTake_t::IsNotTake
+                : IsTake_t::IsTake,
+            IsInitialization_t::IsInitialization);
+        if (ExistentialArgDescriptor[ArgDesc.Index].DestroyAddrUse) {
+          NewFBuilder.createDestroyAddr(InsertLoc, NewArg);
+        } else {
+          MissingDestroyUse = true;
+        }
+        ValueMap.insert(std::make_pair(OrigArg, SILValue(ASI)));
+        break;
+      }
+      case ExistentialRepresentation::Class: {
+        ///  Simple case: Create an init_existential.
+        /// %5 = init_existential_ref %0 : $T : $T, $P
+        auto *InitRef = NewFBuilder.createInitExistentialRef(
+            InsertLoc, ArgDesc.Arg->getType(),
+            NewArg->getType().getASTType()->getCanonicalType(), NewArg,
+            Conformances);
+        ValueMap.insert(std::make_pair(OrigArg, SILValue(InitRef)));
+        break;
+      }
+      default: {
+        llvm_unreachable("Unhandled existential type in ExistentialTransform!");
+        break;
+      }
+      };
+    } else {
+      /// Arguments that are not rewritten.
+      auto Ty = params[ArgDesc.Index].getType();
+      if (!(Ty->hasTypeParameter() || Ty->hasArchetype())) {
+        NewArg = ClonedEntryBB->createFunctionArgument(ArgDesc.Arg->getType(),
+                                                       ArgDesc.Decl);
+        NewArg->setOwnershipKind(ValueOwnershipKind(
+            M, ArgDesc.Arg->getType(), ArgDesc.Arg->getArgumentConvention()));
```

Comment:
I think both branches of this 'if' can be folded into one -- the generic case should subsume the non-generic case